In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
# lectura de datos de tablas fct
fct_precios_insiders_AAPL = pd.read_csv(('../Dataset/clean/AAPL_fct_precios_insider.csv'))
fct_precios_insiders_AAPL.head()

Unnamed: 0  trade_date ticker      insider_name title trade_type  \
0           0  2015-01-23   AAPL  Riccio Daniel J.   SVP          S   
1           1  2015-02-18   AAPL       Jung Andrea   Dir          S   
2           2  2015-03-06   AAPL      Maestri Luca   CFO          S   
3           3  2015-03-09   AAPL      Maestri Luca   CFO          S   
4           4  2015-03-18   AAPL      Maestri Luca   CFO          S   

   quantity_of_shares  owned delta_owned    value   Open  Close  movimiento  
0                3804      0       -100%   428955  24.98  25.13        0.15  
1               40000  14595        -73%  5125200  28.50  28.75        0.25  
2                3400  14124        -19%   437920  28.68  28.27       -0.41  
3                2800  11324        -20%   361116  28.58  28.39       -0.19  
4               10823    501        -96%  1394219  28.36  28.69        0.33

In [3]:
# eliminar el simbolo % y el guion - de la columna "delta_owned".
fct_precios_insiders_AAPL['delta_owned'] = fct_precios_insiders_AAPL['delta_owned'].str.replace('%', '').str.replace('-', '')

In [4]:
fct_precios_insiders_AAPL

Unnamed: 0  trade_date ticker        insider_name title trade_type  \
0             0  2015-01-23   AAPL    Riccio Daniel J.   SVP          S   
1             1  2015-02-18   AAPL         Jung Andrea   Dir          S   
2             2  2015-03-06   AAPL        Maestri Luca   CFO          S   
3             3  2015-03-09   AAPL        Maestri Luca   CFO          S   
4             4  2015-03-18   AAPL        Maestri Luca   CFO          S   
..          ...         ...    ...                 ...   ...        ...   
216         216  2024-10-04   AAPL        Maestri Luca   CFO          S   
217         217  2024-11-15   AAPL   Levinson Arthur D   Dir          S   
218         218  2024-11-18   AAPL         Kondo Chris   PAO          S   
219         219  2024-12-16   AAPL  Williams Jeffrey E   COO          S   
220         220  2025-02-03   AAPL   Levinson Arthur D   Dir          S   

     quantity_of_shares    owned delta_owned     value    Open   Close  \
0                  3804        0         100    428955   24.98   25.13   
1                 40000    14595          73   5125200   28.50   28.75   
2                  3400    14124          19    437920   28.68   28.27   
3                  2800    11324          20    361116   28.58   28.39   
4                 10823      501          96   1394219   28.36   28.69   
..                  ...      ...         ...       ...     ...     ...   
216               59305   107788          35  13433769  227.40  226.30   
217              200000  4215576           5  45464500  226.15  224.75   
218                4130    15419          21    945233  225.00  227.77   
219              100000   389944          20  24997395  247.72  250.76   
220                1516  4215576           0    343147  229.74  227.76   

     movimiento  
0          0.15  
1          0.25  
2         -0.41  
3         -0.19  
4          0.33  
..          ...  
216       -1.10  
217       -1.40  
218        2.77  
219        3.04  
220       -1.98  

[221 rows x 13 columns]

# COMENTARIO
Necesitaba quitar el símbolo de % de la columna delta_owned porque quiero realizar una métrica para saber si el insider vendió una parte significativa de sus acciones. Por ejemplo, lo quiero estimar en un umbral del 50%.

Hay que entender que la columna delta_owned representa el % de acciones que tiene esa persona luego de haber realizado una transacción, por ejemplo, si es venta el delta_owned será una disminución, si es compra será un aumento. 

Para el caso de las personas que su delta_owned es 0, no quiere decir que no hicieron transacciones, quiere decir que el porcentaje negociado es ínfimo frente al total de las acciones que poseen. Esto se puede evidenciar en la línea 220 de la tabla fct.

In [5]:
# transformar la columna "delta_owned" a tipo int 
fct_precios_insiders_AAPL['delta_owned'] = fct_precios_insiders_AAPL['delta_owned'].astype(int)

In [6]:
fct_precios_insiders_AAPL

Unnamed: 0  trade_date ticker        insider_name title trade_type  \
0             0  2015-01-23   AAPL    Riccio Daniel J.   SVP          S   
1             1  2015-02-18   AAPL         Jung Andrea   Dir          S   
2             2  2015-03-06   AAPL        Maestri Luca   CFO          S   
3             3  2015-03-09   AAPL        Maestri Luca   CFO          S   
4             4  2015-03-18   AAPL        Maestri Luca   CFO          S   
..          ...         ...    ...                 ...   ...        ...   
216         216  2024-10-04   AAPL        Maestri Luca   CFO          S   
217         217  2024-11-15   AAPL   Levinson Arthur D   Dir          S   
218         218  2024-11-18   AAPL         Kondo Chris   PAO          S   
219         219  2024-12-16   AAPL  Williams Jeffrey E   COO          S   
220         220  2025-02-03   AAPL   Levinson Arthur D   Dir          S   

     quantity_of_shares    owned  delta_owned     value    Open   Close  \
0                  3804        0          100    428955   24.98   25.13   
1                 40000    14595           73   5125200   28.50   28.75   
2                  3400    14124           19    437920   28.68   28.27   
3                  2800    11324           20    361116   28.58   28.39   
4                 10823      501           96   1394219   28.36   28.69   
..                  ...      ...          ...       ...     ...     ...   
216               59305   107788           35  13433769  227.40  226.30   
217              200000  4215576            5  45464500  226.15  224.75   
218                4130    15419           21    945233  225.00  227.77   
219              100000   389944           20  24997395  247.72  250.76   
220                1516  4215576            0    343147  229.74  227.76   

     movimiento  
0          0.15  
1          0.25  
2         -0.41  
3         -0.19  
4          0.33  
..          ...  
216       -1.10  
217       -1.40  
218        2.77  
219        3.04  
220       -1.98  

[221 rows x 13 columns]

In [7]:
# crear una nueva columna para determinar si el insider ha vendido o comprado en base al valor de la columna "delta_owned" mayor al 10% debe colocar 1 es venta significativa y 0 si no lo es

fct_precios_insiders_AAPL['significant_transaction'] = np.where(fct_precios_insiders_AAPL['delta_owned'] >= 10, 1, 0)
fct_precios_insiders_AAPL.head(10)

Unnamed: 0  trade_date ticker       insider_name title trade_type  \
0           0  2015-01-23   AAPL   Riccio Daniel J.   SVP          S   
1           1  2015-02-18   AAPL        Jung Andrea   Dir          S   
2           2  2015-03-06   AAPL       Maestri Luca   CFO          S   
3           3  2015-03-09   AAPL       Maestri Luca   CFO          S   
4           4  2015-03-18   AAPL       Maestri Luca   CFO          S   
5           5  2015-04-02   AAPL  Ahrendts Angela J   SVP          S   
6           6  2015-04-06   AAPL  Ahrendts Angela J   SVP          S   
7           7  2015-04-16   AAPL       Maestri Luca   CFO          S   
8           8  2015-04-20   AAPL       Maestri Luca   CFO          S   
9           9  2015-04-30   AAPL   Riccio Daniel J.   SVP          S   

   quantity_of_shares  owned  delta_owned    value   Open  Close  movimiento  \
0                3804      0          100   428955  24.98  25.13        0.15   
1               40000  14595           73  5125200  28.50  28.75        0.25   
2                3400  14124           19   437920  28.68  28.27       -0.41   
3                2800  11324           20   361116  28.58  28.39       -0.19   
4               10823    501           96  1394219  28.36  28.69        0.33   
5               25000  99728           20  3119874  27.92  27.99        0.07   
6               44197  55531           44  5607762  27.80  28.44        0.64   
7                 700   5936           11    88788  28.20  28.18       -0.02   
8                5936      0          100   758080  28.04  28.50        0.46   
9               24090  72255           25  3031463  28.73  27.95       -0.78   

   significant_transaction  
0                        1  
1                        1  
2                        1  
3                        1  
4                        1  
5                        1  
6                        1  
7                        1  
8                        1  
9                        1

In [8]:
# quiero filtrar los datos que muestre los 1 en significant_transaction y los valores menores a 0 en movimiento

filtered_data = fct_precios_insiders_AAPL[
	(fct_precios_insiders_AAPL['significant_transaction'] == 1) & 
	(fct_precios_insiders_AAPL['movimiento'] < 0)
]
filtered_data


Unnamed: 0  trade_date ticker        insider_name title trade_type  \
2             2  2015-03-06   AAPL        Maestri Luca   CFO          S   
3             3  2015-03-09   AAPL        Maestri Luca   CFO          S   
7             7  2015-04-16   AAPL        Maestri Luca   CFO          S   
9             9  2015-04-30   AAPL    Riccio Daniel J.   SVP          S   
10           10  2015-05-04   AAPL   Ahrendts Angela J   SVP          S   
15           15  2015-07-24   AAPL    Riccio Daniel J.   SVP          S   
16           16  2015-07-24   AAPL        Wagner Susan   Dir          P   
20           20  2015-10-16   AAPL        Maestri Luca   CFO          S   
23           23  2015-10-30   AAPL       Cue Eduardo H   SVP          S   
40           40  2016-08-02   AAPL         Kondo Chris   PAO          S   
44           44  2016-08-16   AAPL      Sewell D Bruce   SVP          S   
50           50  2016-09-21   AAPL    Riccio Daniel J.   SVP          S   
51           51  2016-10-03   AAPL  Williams Jeffrey E   COO          S   
52           52  2016-10-27   AAPL    Riccio Daniel J.   SVP          S   
53           53  2016-10-27   AAPL   Schiller Philip W   SVP          S   
55           55  2016-11-03   AAPL    Riccio Daniel J.   SVP          S   
74           74  2017-02-16   AAPL     Federighi Craig   SVP          S   
79           79  2017-04-05   AAPL        Maestri Luca   CFO          S   
80           80  2017-04-06   AAPL    Riccio Daniel J.   SVP          S   
81           81  2017-04-19   AAPL        Maestri Luca   CFO          S   
86           86  2017-05-16   AAPL    Riccio Daniel J.   SVP          S   
89           89  2017-05-23   AAPL     Federighi Craig   SVP          S   
93           93  2017-07-06   AAPL    Riccio Daniel J.   SVP          S   
97           97  2017-08-16   AAPL         Kondo Chris   PAO          S   
98           98  2017-08-16   AAPL        Srouji Johny   SVP          S   
104         104  2017-11-14   AAPL    Riccio Daniel J.   SVP          S   
116         116  2018-05-17   AAPL    Riccio Daniel J.   SVP          S   
124         124  2018-08-09   AAPL     Federighi Craig   SVP          S   
129         129  2018-10-04   AAPL        Maestri Luca   CFO          S   
130         130  2018-11-19   AAPL         Kondo Chris   PAO          S   
133         133  2019-05-02   AAPL  Williams Jeffrey E   COO          S   
135         135  2019-08-01   AAPL         Kondo Chris   PAO          S   
139         139  2019-10-02   AAPL  Williams Jeffrey E   COO          S   
146         146  2020-04-07   AAPL        Maestri Luca   CFO          S   
147         147  2020-04-16   AAPL     O'Brien Deirdre   SVP          S   
152         152  2020-10-16   AAPL         Kondo Chris   PAO          S   
153         153  2020-10-16   AAPL     O'Brien Deirdre   SVP          S   
158         158  2021-04-16   AAPL     O'Brien Deirdre   SVP          S   
162         162  2021-08-06   AAPL     O'Brien Deirdre   SVP          S   
163         163  2021-08-25   AAPL      Cook Timothy D   CEO          S   
164         164  2021-10-04   AAPL  Williams Jeffrey E   COO          S   
165         165  2021-10-04   AAPL     O'Brien Deirdre   SVP          S   
184         184  2023-03-22   AAPL  Williams Jeffrey E   COO          S   
188         188  2023-04-04   AAPL  Williams Jeffrey E   COO          S   
192         192  2023-08-07   AAPL     O'Brien Deirdre   SVP          S   
199         199  2023-11-17   AAPL  Adams Katherine L.   SVP          S   
200         200  2023-11-29   AAPL         Kondo Chris   PAO          S   
203         203  2024-04-02   AAPL  Adams Katherine L.   SVP          S   
204         204  2024-04-02   AAPL     O'Brien Deirdre   SVP          S   
216         216  2024-10-04   AAPL        Maestri Luca   CFO          S   

     quantity_of_shares    owned  delta_owned      value    Open   Close  \
2                  3400    14124           19     437920   28.68   28.27   
3                  2800    

In [9]:
filtered_data_2 = fct_precios_insiders_AAPL[
	(fct_precios_insiders_AAPL['significant_transaction'] == 1) & 
	(fct_precios_insiders_AAPL['movimiento'] > 0)
]
filtered_data_2

Unnamed: 0  trade_date ticker        insider_name title trade_type  \
0             0  2015-01-23   AAPL    Riccio Daniel J.   SVP          S   
1             1  2015-02-18   AAPL         Jung Andrea   Dir          S   
4             4  2015-03-18   AAPL        Maestri Luca   CFO          S   
5             5  2015-04-02   AAPL   Ahrendts Angela J   SVP          S   
6             6  2015-04-06   AAPL   Ahrendts Angela J   SVP          S   
..          ...         ...    ...                 ...   ...        ...   
212         212  2024-10-02   AAPL     O'Brien Deirdre   SVP          S   
214         214  2024-10-02   AAPL  Adams Katherine L.   SVP          S   
215         215  2024-10-02   AAPL  Williams Jeffrey E   COO          S   
218         218  2024-11-18   AAPL         Kondo Chris   PAO          S   
219         219  2024-12-16   AAPL  Williams Jeffrey E   COO          S   

     quantity_of_shares   owned  delta_owned     value    Open   Close  \
0                  3804       0          100    428955   24.98   25.13   
1                 40000   14595           73   5125200   28.50   28.75   
4                 10823     501           96   1394219   28.36   28.69   
5                 25000   99728           20   3119874   27.92   27.99   
6                 44197   55531           44   5607762   27.80   28.44   
..                  ...     ...          ...       ...     ...     ...   
212               61019  136572           31  13843382  225.39  226.28   
214               61019  187043           25  13802297  225.39  226.28   
215               59730  489944           11  13550148  225.39  226.28   
218                4130   15419           21    945233  225.00  227.77   
219              100000  389944           20  24997395  247.72  250.76   

     movimiento  significant_transaction  
0          0.15                        1  
1          0.25                        1  
4          0.33                        1  
5          0.07                        1  
6          0.64                        1  
..          ...                      ...  
212        0.89                        1  
214        0.89                        1  
215        0.89                        1  
218        2.77                        1  
219        3.04                        1  

[118 rows x 14 columns]

In [10]:
# crear columna para determinar el impacto negativo si el insider ha vendido en base al valor de movimiento y significant_transaction, que el output sea true o false, con inplace=true
fct_precios_insiders_AAPL['impacto_negativo'] = fct_precios_insiders_AAPL.apply(
    lambda x: True if (x['movimiento'] < 0 and x['significant_transaction'] == 1) else False, axis=1
)

In [11]:
fct_precios_insiders_AAPL

Unnamed: 0  trade_date ticker        insider_name title trade_type  \
0             0  2015-01-23   AAPL    Riccio Daniel J.   SVP          S   
1             1  2015-02-18   AAPL         Jung Andrea   Dir          S   
2             2  2015-03-06   AAPL        Maestri Luca   CFO          S   
3             3  2015-03-09   AAPL        Maestri Luca   CFO          S   
4             4  2015-03-18   AAPL        Maestri Luca   CFO          S   
..          ...         ...    ...                 ...   ...        ...   
216         216  2024-10-04   AAPL        Maestri Luca   CFO          S   
217         217  2024-11-15   AAPL   Levinson Arthur D   Dir          S   
218         218  2024-11-18   AAPL         Kondo Chris   PAO          S   
219         219  2024-12-16   AAPL  Williams Jeffrey E   COO          S   
220         220  2025-02-03   AAPL   Levinson Arthur D   Dir          S   

     quantity_of_shares    owned  delta_owned     value    Open   Close  \
0                  3804        0          100    428955   24.98   25.13   
1                 40000    14595           73   5125200   28.50   28.75   
2                  3400    14124           19    437920   28.68   28.27   
3                  2800    11324           20    361116   28.58   28.39   
4                 10823      501           96   1394219   28.36   28.69   
..                  ...      ...          ...       ...     ...     ...   
216               59305   107788           35  13433769  227.40  226.30   
217              200000  4215576            5  45464500  226.15  224.75   
218                4130    15419           21    945233  225.00  227.77   
219              100000   389944           20  24997395  247.72  250.76   
220                1516  4215576            0    343147  229.74  227.76   

     movimiento  significant_transaction  impacto_negativo  
0          0.15                        1             False  
1          0.25                        1             False  
2         -0.41                        1              True  
3         -0.19                        1              True  
4          0.33                        1             False  
..          ...                      ...               ...  
216       -1.10                        1              True  
217       -1.40                        0             False  
218        2.77                        1             False  
219        3.04                        1             False  
220       -1.98                        0             False  

[221 rows x 15 columns]

In [12]:
fct_precios_insiders_AAPL['año'] = pd.to_datetime(fct_precios_insiders_AAPL['trade_date']).dt.year
fct_precios_insiders_AAPL['mes'] = pd.to_datetime(fct_precios_insiders_AAPL['trade_date']).dt.month

In [13]:
fct_precios_insiders_AAPL

Unnamed: 0  trade_date ticker        insider_name title trade_type  \
0             0  2015-01-23   AAPL    Riccio Daniel J.   SVP          S   
1             1  2015-02-18   AAPL         Jung Andrea   Dir          S   
2             2  2015-03-06   AAPL        Maestri Luca   CFO          S   
3             3  2015-03-09   AAPL        Maestri Luca   CFO          S   
4             4  2015-03-18   AAPL        Maestri Luca   CFO          S   
..          ...         ...    ...                 ...   ...        ...   
216         216  2024-10-04   AAPL        Maestri Luca   CFO          S   
217         217  2024-11-15   AAPL   Levinson Arthur D   Dir          S   
218         218  2024-11-18   AAPL         Kondo Chris   PAO          S   
219         219  2024-12-16   AAPL  Williams Jeffrey E   COO          S   
220         220  2025-02-03   AAPL   Levinson Arthur D   Dir          S   

     quantity_of_shares    owned  delta_owned     value    Open   Close  \
0                  3804        0          100    428955   24.98   25.13   
1                 40000    14595           73   5125200   28.50   28.75   
2                  3400    14124           19    437920   28.68   28.27   
3                  2800    11324           20    361116   28.58   28.39   
4                 10823      501           96   1394219   28.36   28.69   
..                  ...      ...          ...       ...     ...     ...   
216               59305   107788           35  13433769  227.40  226.30   
217              200000  4215576            5  45464500  226.15  224.75   
218                4130    15419           21    945233  225.00  227.77   
219              100000   389944           20  24997395  247.72  250.76   
220                1516  4215576            0    343147  229.74  227.76   

     movimiento  significant_transaction  impacto_negativo   año  mes  
0          0.15                        1             False  2015    1  
1          0.25                        1             False  2015    2  
2         -0.41                        1              True  2015    3  
3         -0.19                        1              True  2015    3  
4          0.33                        1             False  2015    3  
..          ...                      ...               ...   ...  ...  
216       -1.10                        1              True  2024   10  
217       -1.40                        0             False  2024   11  
218        2.77                        1             False  2024   11  
219        3.04                        1             False  2024   12  
220       -1.98                        0             False  2025    2  

[221 rows x 17 columns]

# COMENTARIO
Determiné un umbral del 10% porque tomando en cuenta que son directores de la compañía tienen mucha influencia y un movimiento de acciones por más pequeño que sea debe ser considerado como una alerta. 
Mi análisis será a partir de verificar el impacto con el cambio del precio de la acción en negativo, es decir, si un director que haya vendido por lo menos el 10% de sus acciones afecta al precio de la acción en negativo, se consideraría un resultado positivo.

In [14]:
# tomar el dataframe limpio y guardarlo como un nuevo csv
fct_precios_insiders_AAPL.to_csv(('../Dataset/ml_dataset/AAPL_ml_dataset.csv'))